### PR TITLE
refactor: merge to use logical plans

### LINKED ISSUE #1753: Writing with large arrow types in MERGE
# Description
It seems it's not possible to write with large arrow types in `.merge()` yet, however there is write support with `write_deltalake(large_types=True)`, we should add this also in merge.

**Use Case**
Converting Polars dataframes to arrow and then merging immediately instead of casting to normal arrow types which may not fit if the arrays are too large.

```
DeltaError: Generic DeltaTable error: Execution error: Fail to build join indices in NestedLoopJoinExec, error:Arrow error: Invalid argument error: Invalid comparison operation: LargeUtf8 == Utf8
```

### LINKED ISSUE #1790: rust merge error - datafusion panics
# Environment

**Delta-rs version**:
0.16.2
**Binding**: 
rust

***
# Bug

**What happened**:
Code panics when trying to merge datafusion's DataFrame into the DeltaTable (on MinIO S3).
Could it be connected to https://github.com/delta-io/delta-rs/issues/1699?

**What you expected to happen**:
Not panic, originally I had INSERT clauses as well, but removed them for example simplification's case as the error was exactly the same.

**How to reproduce it**:
I've made a minimal example here: https://github.com/PiotrGierda/minimal_example_deltalake_datafusion
Need to provide valid MinIO credentials and details as ENVVARs when running with `cargo run`
https://github.com/PiotrGierda/minimal_example_deltalake_datafusion/blob/main/src/main.rs#L99C5-L102C25:
```rust
let (_table, _metrics) = DeltaOps(target_table)
        .merge(source_df, col("__id").eq(col("source.__id")))
        .with_source_alias("source")
        .await.unwrap();
```
results in:
```
     Running `target/debug/minimal_example`
+--------+-------------+-------------+
| __id   | __createdat | __updatedat |
+--------+-------------+-------------+
| csvboi | 1698670149  | 1698670149  |
+--------+-------------+-------------+
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Generic("Internal error: CoalescePartitionsExec requires at least one input partition\n\nbacktrace:    0: datafusion_common::error::DataFusionError::get_back_trace\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-common-31.0.0/src/error.rs:423:26\n   1: <datafusion::physical_plan::coalesce_partitions::CoalescePartitionsExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-31.0.0/src/physical_plan/coalesce_partitions.rs:128:18\n   2: <datafusion::physical_plan::joins::nested_loop_join::NestedLoopJoinExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-31.0.0/src/physical_plan/joins/nested_loop_join.rs:250:31\n   3: <datafusion::physical_plan::projection::ProjectionExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-31.0.0/src/physical_plan/projection.rs:326:20\n   4: <datafusion::physical_plan::projection::ProjectionExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-31.0.0/src/physical_plan/projection.rs:326:20\n   5: <deltalake::operations::datafusion_utils::MetricObserverExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deltalake-0.16.2/src/operations/mod.rs:324:23\n   6: <datafusion::physical_plan::filter::FilterExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-31.0.0/src/physical_plan/filter.rs:179:20\n   7: <datafusion::physical_plan::projection::ProjectionExec as datafusion::physical_plan::ExecutionPlan>::execute\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/datafusion-31.0.0/src/physical_plan/projection.rs:326:20\n   8: deltalake::operations::write::write_execution_plan::{{closure}}\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deltalake-0.16.2/src/operations/write.rs:279:26\n   9: deltalake::operations::merge::execute::{{closure}}\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deltalake-0.16.2/src/operations/merge.rs:990:6\n  10: <deltalake::operations::merge::MergeBuilder as core::future::into_future::IntoFuture>::into_future::{{closure}}\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/deltalake-0.16.2/src/operations/merge.rs:1087:14\n  11: <core::pin::Pin<P> as core::future::future::Future>::poll\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/future/future.rs:125:9\n  12: minimal_example::main::{{closure}}\n             at ./src/main.rs:102:10\n  13: tokio::runtime::park::CachedParkThread::block_on::{{closure}}\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/park.rs:282:63\n  14: tokio::runtime::coop::with_budget\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/coop.rs:107:5\n  15: tokio::runtime::coop::budget\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/coop.rs:73:5\n  16: tokio::runtime::park::CachedParkThread::block_on\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/park.rs:282:31\n  17: tokio::runtime::context::blocking::BlockingRegionGuard::block_on\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/context/blocking.rs:66:9\n  18: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/scheduler/multi_thread/mod.rs:87:13\n  19: tokio::runtime::context::runtime::enter_runtime\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/context/runtime.rs:65:16\n  20: tokio::runtime::scheduler::multi_thread::MultiThread::block_on\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/scheduler/multi_thread/mod.rs:86:9\n  21: tokio::runtime::runtime::Runtime::block_on\n             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/runtime.rs:350:45\n  22: minimal_example::main\n             at ./src/main.rs:99:5\n  23: core::ops::function::FnOnce::call_once\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5\n  24: std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/sys_common/backtrace.rs:135:18\n  25: std::rt::lang_start::{{closure}}\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/rt.rs:166:18\n  26: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:284:13\n  27: std::panicking::try::do_call\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:500:40\n  28: std::panicking::try\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:464:19\n  29: std::panic::catch_unwind\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panic.rs:142:14\n  30: std::rt::lang_start_internal::{{closure}}\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/rt.rs:148:48\n  31: std::panicking::try::do_call\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:500:40\n  32: std::panicking::try\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:464:19\n  33: std::panic::catch_unwind\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panic.rs:142:14\n  34: std::rt::lang_start_internal\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/rt.rs:148:20\n  35: std::rt::lang_start\n             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/rt.rs:165:17\n  36: main\n  37: __libc_start_call_main\n             at ./csu/../sysdeps/nptl/libc_start_call_main.h:58:16\n  38: __libc_start_main_impl\n             at ./csu/../csu/libc-start.c:392:3\n  39: _start\n.\nThis was likely caused by a bug in DataFusion's code and we would welcome that you file an bug report in our issue tracker")', src/main.rs:102:16
stack backtrace:
   0: rust_begin_unwind
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/std/src/panicking.rs:593:5
   1: core::panicking::panic_fmt
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/panicking.rs:67:14
   2: core::result::unwrap_failed
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/result.rs:1651:5
   3: core::result::Result<T,E>::unwrap
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/result.rs:1076:23
   4: minimal_example::main::{{closure}}
             at ./src/main.rs:99:30
   5: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/park.rs:282:63
   6: tokio::runtime::coop::with_budget
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/coop.rs:107:5
   7: tokio::runtime::coop::budget
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/coop.rs:73:5
   8: tokio::runtime::park::CachedParkThread::block_on
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/park.rs:282:31
   9: tokio::runtime::context::blocking::BlockingRegionGuard::block_on
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/context/blocking.rs:66:9
  10: tokio::runtime::scheduler::multi_thread::MultiThread::block_on::{{closure}}
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/scheduler/multi_thread/mod.rs:87:13
  11: tokio::runtime::context::runtime::enter_runtime
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/context/runtime.rs:65:16
  12: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/scheduler/multi_thread/mod.rs:86:9
  13: tokio::runtime::runtime::Runtime::block_on
             at /home/piotr/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.33.0/src/runtime/runtime.rs:350:45
  14: minimal_example::main
             at ./src/main.rs:99:5
  15: core::ops::function::FnOnce::call_once
             at /rustc/5680fa18feaa87f3ff04063800aec256c3d4b4be/library/core/src/ops/function.rs:250:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

**More details**:
The same happens with INSERT clauses and with some data already in the table.

### PR DESCRIPTION
**Pull request by Blajda**
# Description
This refactors the merge operation to use DataFusion's DataFrame and LogicalPlan APIs

The NLJ is eliminated and the query planner can pick the optimal join operator. This also enables the operation to use multiple threads and should result in significant speed up.
Merge is still limited to using a single thread in some area. When collecting benchmarks, I encountered multiple OoM issues with Datafusion's hash join implementation. There are multiple tickets upstream open regarding this. For now, I've limited the number of partitions to just 1 to prevent this.

Predicates passed as SQL are also easier to use now. Manual casting was required to ensure data types were aligned. Now the logical plan will perform type coercion when optimizing the plan. 

# Related Issues
- enhances #850
- closes #1790 
- closes #1753

### CODE REVIEW

**Comment by Blajda** [comment]
Can be merged after #1639

**Comment by ion-elgreco** [comment]
@Blajda is this ready to be merged now? The issue and PR you linked have been merged/resolved now

**Comment by Blajda** [comment]
@ion-elgreco It depends on a new version of datafusion being released and then having  https://github.com/delta-io/delta-rs/pull/1775 updated and merged 
I also need to make one small change to prevent predicate pushdowns. The benchmark code is done I'll raise a PR for that soon.

**Comment by ion-elgreco** [comment]
@Blajda oh nice! Really curious to see how fast the rust merge is versus the scala merge 👀

**Comment by Blajda** [comment]
Below are the benchmark results. These were executed with the standard merge benchmarks compiled with release flags. 
These are the specs of my desktop for some context;

CPU: AMD Ryzen 7 5800X 8-Core Processor
Memory: 32gb @ 2133 mhz

Writes were done to a HDD.

```
+-------------------------------------------------------------------------------------+---------------------+--------------------+------------------------------------------+
| name                                                                                | before_duration_avg | after_duration_avg | before_duration_avg / after_duration_avg |
+-------------------------------------------------------------------------------------+---------------------+--------------------+------------------------------------------+
| delete_only_fileMatchedFraction_0.05_rowMatchedFraction_0.05                        | 16429.0             | 6298.0             | 2.6086059066370275                       |
| multiple_insert_only_fileMatchedFraction_0.05_rowNotMatchedFraction_0.05            | 14731.0             | 5436.0             | 2.709896983075791                        |
| multiple_insert_only_fileMatchedFraction_0.05_rowNotMatchedFraction_0.50            | 71219.0             | 5260.0             | 13.539733840304182                       |
| multiple_insert_only_fileMatchedFraction_0.05_rowNotMatchedFraction_1.0             | 116833.0            | 5336.0             | 21.89523988005997                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.01_rowNotMatchedFraction_0.1   | 22406.0             | 5003.0             | 4.478512892264641                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.0_rowNotMatchedFraction_0.1    | 34476.0             | 4496.0             | 7.668149466192171                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.1_rowNotMatchedFraction_0.0    | 22723.0             | 4841.0             | 4.693864903945466                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.1_rowNotMatchedFraction_0.01   | 21223.0             | 4897.0             | 4.333877884419032                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.5_rowNotMatchedFraction_0.001  | 68764.0             | 4673.0             | 14.715172266210143                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.99_rowNotMatchedFraction_0.001 | 113265.0            | 4676.0             | 24.22262617621899                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_1.0_rowNotMatchedFraction_0.001  | 91681.0             | 4883.0             | 18.77554781896375                        |
| upsert_fileMatchedFraction_0.5_rowMatchedFraction_0.001_rowNotMatchedFraction_0.001 | 12294.0             | 4685.0             | 2.624119530416222                        |
| upsert_fileMatchedFraction_1.0_rowMatchedFraction_0.001_rowNotMatchedFraction_0.001 | 14956.0             | 4868.0             | 3.0723089564502875                       |
+-------------------------------------------------------------------------------------+---------------------+--------------------+------------------------------------------+
```

**Comment by ion-elgreco** [comment]
@Blajda I assume after_duration is after logical plans. 

Looks quite a speed bump and also more consistent across the different queries, so something else seems to be a limiting there? Maybe your HDD? 😛 

What's the scale btw, milliseconds?

**Comment by ion-elgreco** [comment]
@Blajda it will also close this one then: https://github.com/delta-io/delta-rs/issues/1753

**Comment by Blajda** [comment]
Before: Merge logical to HDD
After Merge Logical to SDD

Currently the size of the write is not the limiting factor 

```
| name                                                                                | before_duration_avg | after_duration_avg | before_duration_avg / after_duration_avg |
+-------------------------------------------------------------------------------------+---------------------+--------------------+------------------------------------------+
| delete_only_fileMatchedFraction_0.05_rowMatchedFraction_0.05                        | 6298.0              | 6089.0             | 1.0343241911643948                       |
| multiple_insert_only_fileMatchedFraction_0.05_rowNotMatchedFraction_0.05            | 5436.0              | 5175.0             | 1.0504347826086957                       |
| multiple_insert_only_fileMatchedFraction_0.05_rowNotMatchedFraction_0.50            | 5260.0              | 5458.0             | 0.9637229754488824                       |
| multiple_insert_only_fileMatchedFraction_0.05_rowNotMatchedFraction_1.0             | 5336.0              | 5343.0             | 0.9986898746022833                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.01_rowNotMatchedFraction_0.1   | 5003.0              | 5035.0             | 0.9936444885799404                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.0_rowNotMatchedFraction_0.1    | 4496.0              | 4165.0             | 1.079471788715486                        |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.1_rowNotMatchedFraction_0.0    | 4841.0              | 4932.0             | 0.9815490673154906                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.1_rowNotMatchedFraction_0.01   | 4897.0              | 4610.0             | 1.0622559652928416                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.5_rowNotMatchedFraction_0.001  | 4673.0              | 4843.0             | 0.9648977906256453                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_0.99_rowNotMatchedFraction_0.001 | 4676.0              | 4481.0             | 1.0435170720821245                       |
| upsert_fileMatchedFraction_0.05_rowMatchedFraction_1.0_rowNotMatchedFraction_0.001  | 4883.0              | 4865.0             | 1.003699897225077                        |
| upsert_fileMatchedFraction_0.5_rowMatchedFraction_0.001_rowNotMatchedFraction_0.001 | 4685.0              | 4393.0             | 1.0664693831094925                       |
| upsert_fileMatchedFraction_1.0_rowMatchedFraction_0.001_rowNotMatchedFraction_0.001 | 4868.0              | 4627.0             | 1.0520855846120596                       |
+-------------------------------------------------------------------------------------+---------------------+--------------------+------------------------------------------+

File: crates/deltalake-core/src/operations/merge.rs
Comment (roeap) [inline crates/deltalake-core/src/operations/merge.rs:1103]: Eventually we should consolidate this into a `DeltaPlanner`, some early experiments are in the deltalake-sql crate, but that is not helpful right now 😆 - maybe add a comment?

File: crates/deltalake-core/src/operations/merge.rs
Comment (wjones127) [inline crates/deltalake-core/src/operations/merge.rs:1116]: @Blajda Why don't you configure a [Greedy memory pool](https://docs.rs/datafusion/latest/datafusion/execution/memory_pool/struct.GreedyMemoryPool.html) in the [RuntimeConfig](https://docs.rs/datafusion/latest/datafusion/execution/runtime_env/struct.RuntimeEnv.html)? Right now the `SessionContext` seems to be configured without a memory limit, so it's no surprise that it can OOM, right?

You can pass that in with [SessionContext::new_with_config_rt](https://docs.rs/datafusion/latest/datafusion/execution/context/struct.SessionContext.html#method.new_with_config_rt)

**Review by wjones127 — APPROVED** [review]
This is great work. I feel like I'm learning a lot about DataFusion just by reviewing your PRs :)

**Review by roeap — COMMENTED** [review]
Looking great and really excited to see us moving towards more proper planning!

As this is quite complex to reason about, should we maybe add some test to validate generated plans via their string representation, much like datafusion does?

That would also help me understand how the metrics observers impact pushdown. I guess since we optimize the input plans before we preserve file skipping etc, but would help to see it :).

**Review by roeap — DISMISSED** [review]

**Comment by Blajda** [comment]
> That would also help me understand how the metrics observers impact pushdown. I guess since we optimize the input plans before we preserve file skipping etc, but would help to see it :).

Merge doesn't support file skipping currently. A new operator that tracks which files have modifications needs to be built and another operator that determine enables push downs depending on which operators are used. (I.e upserts can support push down but anything with a delete would require additional work)


> As this is quite complex to reason about, should we maybe add some test to validate generated plans via their string representation, much like datafusion does?

I think this would be possible but it would require additional refactoring. Current the string repr is just spaghetti of a bunch of case statements and parquet read nodes.

**Comment by roeap** [comment]
Sounds like a follow-up to me...

File: crates/deltalake-core/src/operations/merge.rs
Comment (Blajda) [reply crates/deltalake-core/src/operations/merge.rs:1116]: I'll give this a try. Issues on the DF tracker suggest that even with a memory pool configured there are allocations that are made outside of the pool. As a start I think limiting memory to 80% of the user's system would be a good start.

Comment (Blajda) [reply crates/deltalake-core/src/operations/merge.rs:1116]: Enabled the mem pool and removed the cpu limit. Still leads to low memory system. I think this will require a follow up on how to tune this best.

**Review by roeap — APPROVED** [review]

**Pull request merged**
